Add testing for DataLoader

AndyDai-nv opened the pull request:
Add one test case for dataloader to improve coverage.

nv-hwoo commented on src/c++/perf_analyzer/perf_utils.h:
```diff
@@ -56,7 +56,7 @@ constexpr uint64_t NANOS_PER_MILLIS = 1000000;
 // Will use the characters specified here to construct random strings
 std::string const character_set =
-    "abcdefghijklmnaoqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890 .?!";
+    "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890 .?!";
```
good catch ;)

nv-hwoo commented on src/c++/perf_analyzer/test_dataloader.cc:
```diff
@@ -104,6 +105,127 @@ TEST_CASE("dataloader: GetTotalSteps")
+
+    cb::Error status = dataloader.ReadDataFromJSON(inputs, outputs, json_file);
+    std::filesystem::remove(json_file);
+    std::cout << status.Message();
```
Is `std::cout` needed here?

nv-hwoo reviewed (commented):
Great work Andy!  The test looks very clean and nice :) Thanks for working on it. I've added just a few comments but otherwise LGTM.

dyastremsky commented:
Clean work on this! Appreciate how comprehensive this is. Please don't forget to update the description.

My concern with this approach is that it's not proper unit testing. We already can't parallelize a few of the Perf Analyzer tests successfully, but with a proper unit test, we should be able to. These tests move us further from that goal, because they actually create/delete the files instead of using mocks. That's better than not having these tests and faster to develop. However, this could cause headaches later.

If we're not going to mock the file/dir to avoid having side effects in these tests, I would leave a ticket in the backlog about doing moving these to proper unit tests later. And also just generally moving to a place where these can be run in parallel with one command, similar to [what we added](https://github.com/triton-inference-server/model_analyzer/blob/11f8a81bc4b717fda4dcfc41c1bf1bdf80be6cd4/.github/workflows/python-package.yaml#L59) to Model Analyzer. That would be a high-value addition.

If we're sticking with this approach, I'd also want to confirm that we have error handling (e.g. setup/teardown or try-catch if that's not possible) that if something goes wrong during the unit test, extra files or directories don't remain on the machine.

CC: @matthewkotila

AndyDai-nv commented:
> Clean work on this! Appreciate how comprehensive this is. Please don't forget to update the description.
> 
> My concern with this approach is that it's not proper unit testing. This is a step in the right direction. However, we already can't parallelize a few of the Perf Analyzer tests successfully, but with a proper unit test, we should be able to. These tests move us further from that goal, because they actually create/delete the files instead of using mocks. That's better than not having these tests and faster to develop. However, this could cause headaches later.
> 
> If we're not going to mock the file/dir to avoid having side effects in these tests, I would leave a ticket in the backlog about doing moving these to proper unit tests later. And also just generally moving to a place where these can be run in parallel with one command, similar to [what we added](https://github.com/triton-inference-server/model_analyzer/blob/11f8a81bc4b717fda4dcfc41c1bf1bdf80be6cd4/.github/workflows/python-package.yaml#L59) to Model Analyzer. That would be a high-value addition.
> 
> If we're sticking with this approach, I'd also want to confirm that we have error handling (e.g. setup/teardown or try-catch if that's not possible) that if something goes wrong during the unit test, extra files or directories don't remain on the machine.
> 
> CC: @matthewkotila



> Clean work on this! Appreciate how comprehensive this is. Please don't forget to update the description.
> 
> My concern with this approach is that it's not proper unit testing. This is a step in the right direction. However, we already can't parallelize a few of the Perf Analyzer tests successfully, but with a proper unit test, we should be able to. These tests move us further from that goal, because they actually create/delete the files instead of using mocks. That's better than not having these tests and faster to develop. However, this could cause headaches later.
> 
> If we're not going to mock the file/dir to avoid having side effects in these tests, I would leave a ticket in the backlog about doing moving these to proper unit tests later. And also just generally moving to a place where these can be run in parallel with one command, similar to [what we added](https://github.com/triton-inference-server/model_analyzer/blob/11f8a81bc4b717fda4dcfc41c1bf1bdf80be6cd4/.github/workflows/python-package.yaml#L59) to Model Analyzer. That would be a high-value addition.
> 
> If we're sticking with this approach, I'd also want to confirm that we have error handling (e.g. setup/teardown or try-catch if that's not possible) that if something goes wrong during the unit test, extra files or directories don't remain on the machine.
> 
> CC: @matthewkotila

Thanks, David. Totally understand the concerns. We updated the JIRA ticket to make this just a draft version before changing to a mock way. And this one can be paused for now.

dyastremsky reviewed (approved):
Great work!

dyastremsky commented:
@AndyDai-nv @matthewkotila Would it be possible to add this as a temporary solution? I don't know when we'll get to the JIRA ticket. I suspect it'll end up getting pushed aside for a while due to other work.

matthewkotila commented:
> @dyastremsky: @AndyDai-nv @matthewkotila Would it be possible to add this as a temporary solution? I don't know when we'll get to the JIRA ticket. I suspect it'll end up getting pushed aside for a while due to other work.

Sure 👍🙏
